src/NHSD.BuyingCatalogue.Solutions.Application/Persistence/PriceReader.cs

```diff
@@ -0,0 +1,83 @@
+        {
+            var prices = await _priceRepository.GetPricesBySolutionIdQueryAsync(solutionId, cancellationToken);
+            Dictionary<int, CataloguePriceBase> dictionary = new Dictionary<int, CataloguePriceBase>();
+
+            foreach (var price in prices)
+            {
+                var cataloguePriceType = Enumerator.FromValue<CataloguePriceType>(price.CataloguePriceTypeId);
+
+                if (Equals(cataloguePriceType, CataloguePriceType.Flat))
+                {
+                    dictionary.Add(price.CataloguePriceId, new CataloguePriceFlat
+                    {
+                        CataloguePriceId = price.CataloguePriceId,
+                        CatalogueItemName = price.CatalogueItemName,
+                        CatalogueItemId = price.CatalogueItemId,
+                        PricingUnit = new PricingUnit
+                        {
+                            Description = price.PricingUnitDescription,
+                            Name = price.PricingUnitName,
+                            TierName = price.PricingUnitTierName
+                        },
+                        TimeUnit = price.TimeUnitId == 0 ? null : Enumerator.FromValue<TimeUnit>(price.TimeUnitId),
+                        CurrencyCode = price.CurrencyCode,
+                        Price = price.FlatPrice.GetValueOrDefault()
+                    });
+                }
+                else if (Equals(cataloguePriceType, CataloguePriceType.Tiered))
+                {
+                    CataloguePriceTier tier;
+
+                    if (dictionary.ContainsKey(price.CataloguePriceId))
+                    {
+                        tier = dictionary[price.CataloguePriceId] as CataloguePriceTier;
+                    }
+
+                    else
+                    {
+                        tier = new CataloguePriceTier
+                        {
+                            CataloguePriceId = price.CataloguePriceId,
+                            CatalogueItemName = price.CatalogueItemName,
+                            CatalogueItemId = price.CatalogueItemId,
+                            CurrencyCode = price.CurrencyCode,
+                            PricingUnit = new PricingUnit
+                            {
+                                Name = price.PricingUnitName,
+                                Description = price.PricingUnitDescription,
+                                TierName = price.PricingUnitTierName
+                            },
+                            TimeUnit = Enumerator.FromValue<TimeUnit>(price.TimeUnitId)
+                        };
+
+                        dictionary.Add(price.CataloguePriceId, tier);
+                    }
+
+                    tier?.TieredPrices.Add(new TieredPrice(price.BandStart.GetValueOrDefault(), price.BandEnd,
+                        price.TieredPrice.GetValueOrDefault()));
+                }
+            }
```

Comment:
I think a standard `enum` could do everything that the 'enumerator' type is being used for here. However, in this case the enum value is being used to drive behaviour – I think we should try to avoid that if possible.

I think it might be better to consider something like the strategy pattern here. Instead of having a set of `if/else` statements we might have a `FlatPriceReader` and a `TieredPriceReader` (or something similar) that encapsulate the logic for reading each of the price types. We could possibly introduce a factory to initialize the relevant reader depending on the price type.

Also, do we actually need the price type? Won't the tiered price be `NULL` if it's a flat pricing structure?

Reply:
Tech Debt Task made: https://buyingcatalog.visualstudio.com/Buying%20Catalogue/_sprints/taskboard/Buying%20Catalogue%20Tech%20Cell/Buying%20Catalogue/Iteration%2027%20Stretch%20Goals?workitem=8004